### PR TITLE
[Core][KV events] Report prefix-cache-reused blocks in full report mode

### PR DESCRIPTION
**Pull request by GongLei-HW**
## Motivation

vLLM's KV cache event publisher (`enable_kv_cache_events`) emits `BlockStored` from `cache_full_blocks()` — i.e. only for **newly allocated** blocks. Blocks **reused from the prefix cache** are never reported, because they are not freshly allocated. External consumers that build a global prefix tree for KV-aware routing (e.g. Dynamo, llm-d, custom gateways) therefore lose visibility into reused prefixes: their mirror trie degrades over time and prefix-match depth collapses to zero.

This adds an opt-in `kv_cache_report_mode` (`"full"`) that *also* emits `BlockStored` events for prefix-cache-reused blocks, so external consumers can observe the full prefix chain of a request.

This is an incremental extension of the KV-events mechanism from RFC #16669; it does not duplicate any existing PR.

## Changes

- `vllm/v1/request.py`: read `kv_cache_report_mode` from `sampling_params.extra_args` (default `"incremental"`).
- `vllm/v1/core/kv_cache_manager.py`: in `get_computed_blocks()`, when the mode is `"full"` and there are cache hits, call `block_pool.emit_cached_block_events()` for the reused blocks. Runs once per request at first schedule (`num_computed_tokens == 0`).
- `vllm/v1/core/block_pool.py`: `emit_cached_block_events()` emits one `BlockStored` for already-cached blocks without mutating block state, mirroring `cache_full_blocks()` (including `group_idx` and `extra_keys`); logged at debug level.

## Usage

Set per-request via the OpenAI-compatible `vllm_xargs` field:

```json
{ "vllm_xargs": { "kv_cache_report_mode": "full" } }
```

## Testing

```
.venv/bin/python -m pytest tests/v1/core/test_prefix_caching.py -k "emit_cached_block_events" -v
```

New unit tests cover: (1) one `BlockStored` with correct `group_idx`/`parent_block_hash`/`token_ids` and no block-state mutation; (2) no events when `enable_kv_cache_events=False`; (3) no events when `num_cached_blocks=0`.

## Out of scope

- Preempt/reschedule may re-trigger emission for the same request; consumers should treat events idempotently (set semantics). Not addressed here.

## AI Assistance Disclosure

This PR was developed with assistance from Claude (Anthropic); in particular, the unit tests for `emit_cached_block_events` were drafted with AI help. All code was reviewed, understood, and validated by the author, and the tests were run locally (`tests/v1/core/test_prefix_caching.py -k emit_cached_block_events`). The corresponding commit carries the `Co-authored-by: Claude` trailer.

### CODE REVIEW

File: vllm/v1/core/block_pool.py [hunk 1]
```diff
@@ -330,6 +330,86 @@ def cache_full_blocks(
...
+            kv_cache_group_id: The KV cache group ID.
+        """
+        if not self.enable_kv_cache_events or num_cached_blocks == 0:
+            return
+
+        # Resolve block hashes (same logic as cache_full_blocks)
+        if block_size == self.hash_block_size:
+            block_hashes: BlockHashList = request.block_hashes
+        else:
+            block_hashes = BlockHashListWithBlockSize(
```
Comment (MengqingCao) [inline vllm/v1/core/block_pool.py]: It seems the check on `block_size` and `hash_block_size` is missed here.

File: vllm/v1/core/block_pool.py [hunk 2]
```diff
@@ -330,6 +330,86 @@ def cache_full_blocks(
...
+            extra_keys, curr_mm_idx = generate_block_hash_extra_keys(
+                request, block_start, block_end, curr_mm_idx
+            )
+            extra_keys_list.append(extra_keys)
+
+        if not cached_hashes:
+            return
+
+        # Reused blocks start from block 0, so parent is None
+        parent_block_hash: ExternalBlockHash | None = None
```
Comment (MengqingCao) [inline vllm/v1/core/block_pool.py]: QQ: Does reused blocks start from block 0 refers to we should reuse the block form the first one of the request blocks, which means no partical cache hit? If so, I think we should optimize the comment.

File: vllm/v1/core/block_pool.py [hunk 3]
```diff
@@ -330,6 +330,86 @@ def cache_full_blocks(
...
+        start_token_idx = 0
+        end_token_idx = num_cached_blocks * block_size
+
+        logger.debug(
+            "EmitCachedBlock event: block_size=%d, "
+            "num_cached_blocks=%d, parent_block_hash=%s, "
+            "token_ids_len=%d, group_idx=%s",
+            block_size,
+            num_cached_blocks,
+            parent_block_hash,
```
Comment (MengqingCao) [inline vllm/v1/core/block_pool.py:423]: no need to report `parent_block_hash` if it is always None?

**Review by MengqingCao — COMMENTED** [review]
Overall I think this is a very useful feature for developers to check the prefix cache blocks event. I only have some tiny questions as illustrated bellow. And I find maybe we could extract the common part of `emit_cached_block_events` and the report section of `cache_full_blocks`, thus the code will be cleaner.

File: vllm/v1/core/block_pool.py (hunk 1 above)
Comment (GongLei-HW) [reply vllm/v1/core/block_pool.py]: Fixed. The missing assert now lives in a shared _resolve_block_hashes helper, and I also extracted _build_block_stored_event — so both cache_full_blocks and emit_cached_block_events share the same block-hash resolution and BlockStored construction, as you suggested. cache_full_blocks behavior is unchanged. Verified with pytest tests/v1/core/test_prefix_caching.py -k emit_cached_block_events → 3 passed.

File: vllm/v1/core/block_pool.py (hunk 2 above)
Comment (GongLei-HW) [reply vllm/v1/core/block_pool.py]: Updated. The comment now clarifies that prefix-cache hits always form a contiguous prefix starting at block 0, so the reused group's parent block hash is None — not "no partial hit", just a prefix that starts from the root.

File: vllm/v1/core/block_pool.py (hunk 3 above)
Comment (GongLei-HW) [reply vllm/v1/core/block_pool.py:423]: Intentional, kept as None. Reused prefix blocks begin at block 0 (the root), so the parent is None by definition. Downstream consumers (e.g. the gateway) rely on the uniform BlockStored shape where parent_block_hash=None denotes a root block — dropping the field would break their parent-chain reconstruction.

**Review by MengqingCao — COMMENTED** [review]
Hi @ivanium , this pr add a event report feature for prefix cache reused blocks, I've made a initial review and it LGTM now, could you please take a look at this pr?

File: vllm/v1/core/block_pool.py [hunk 4]
```diff
@@ -338,23 +328,137 @@ def cache_full_blocks(
...
+        ``block_size`` granularity (``block_size`` must be a multiple of
+        ``hash_block_size``, which happens when KV cache groups differ in
+        block size).
+        """
+        if block_size == self.hash_block_size:
+            return request.block_hashes
+        assert block_size % self.hash_block_size == 0
+        return BlockHashListWithBlockSize(
+            request.block_hashes, self.hash_block_size, block_size
+        )
```
Comment (ivanium) [inline vllm/v1/core/block_pool.py]: Perhaps we can make this a helper func in `kv_cache_utils.py` for more general usage.

**Review by ivanium — COMMENTED** [review]
Thanks @GongLei-HW for the PR and thanks @MengqingCao for the review! LGTM overall; left one nit; cc @njhill and @zhewenl for a second look.

File: vllm/v1/core/block_pool.py (hunk 4 above)
Comment (GongLei-HW) [reply vllm/v1/core/block_pool.py]: Done in 7adb8850. Lifted _resolve_block_hashes to a module-level resolve_block_hashes(request, hash_block_size, block_size) in kv_cache_utils.py, right after the BlockHashList alias, so it's reusable outside BlockPool. cache_full_blocks and emit_cached_block_events now both call the shared helper. Behavior unchanged; tests/v1/core/test_prefix_caching.py -k emit_cached_block_events passes.
Thanks @ivanium! Addressed in 7adb8850. Pinging @njhill @zhewenl for a second look when you have a moment.

**Comment by zhewenl** [comment]
looks good, but I wonder if we want to add some documentation about kv events in the future(like which connector has kv events supported; and what's advanced usage of kv events, like this PR and https://github.com/vllm-project/vllm/pull/43468)

**Comment by GongLei-HW** [comment]
@zhewenl agreed, a dedicated KV events docs section would be valuable. Right now the events are only touched in `docs/features/kv_offloading_usage.md` and `docs/design/prefix_caching.md`, with no overview of event types or which connectors consume them.

I'd keep this PR focused on the feature and do the docs as a follow-up — the "which connector supports KV events" part needs a survey across Dynamo/llm-d/NIXL that's better scoped as its own effort. Happy to open a follow-up PR extending `kv_offloading_usage.md` or a new `docs/features/kv_events.md`. Prefer that now or after this merges?

**Comment by zhewenl** [comment]
sounds good, a follow up PR makes sense!

**Comment by GongLei-HW** [comment]
The CI failures look unrelated to this change. In `amd-entrypoints-integration-api-server-openai`, the only 2 failing tests are audio-transcription (Whisper) tests:

```
FAILED entrypoints/openai/test_run_batch.py::test_transcription
FAILED entrypoints/openai/test_run_batch.py::test_transcription_http_url
= 2 failed, 215 passed
```

These exercise the audio batch API path, not anything touched here (`block_pool.py` / `kv_cache_utils.py`). The NVIDIA equivalent `entrypoints-integration-api-server` passes, as do the KV/prefix-cache suites on both platforms (`v1-core-plus-kv-plus-metrics`, `amd-v1-core-plus-kv-plus-metrics-mi325-1`). The other AMD jobs (`amd-multi-modal-...-whis`, `amd-v1-spec-decode-mi300-1`, `amd-v1-sample-plus-logits-mi325-1`) and `docker-build-cpu-image` likewise don't touch this code path.

I don't have permission to retry the Buildkite jobs — could a maintainer retry the failed ones? Happy to address anything if a real issue is found. Thanks!

**Comment by GongLei-HW** [comment]
Looked at the two CPU failures — both are unrelated to this change:

- **`multi-modal-processor-cpu`**: the single failure is `models/multimodal/processing/test_common.py::test_processing_correctness[...baidu/ERNIE-4.5-VL-28B-A3B-PT]` — a `requests.exceptions.ConnectTimeout` downloading `Roboto-Regular.ttf` from `paddlenlp.bj.bcebos.com` (a font file fetch; pure network flake).
- **`async-engine-inputs-utils-worker-config-cpu`**: the single failure is `tool_parsers/test_structural_tag_registry.py::test_get_model_structural_tag_supports_vllm_hermes` — a Hermes tool-call structural-tag schema assertion in `tool_parsers/`, which doesn't touch `block_pool.py`/`kv_cache_utils.py` (likely a main regression from another PR).

The 4 AMD GPU jobs remain red but `amd-v1-core-plus-kv-plus-metrics-mi325-1` is green on the same build. The KV/prefix-cache suites pass on both platforms (`v1-core-plus-kv-plus-metrics`, `amd-v1-core-plus-kv-plus-metrics-mi325-1`). Could a maintainer retry the flaky jobs? Thanks!

**Comment by GongLei-HW** [comment]
All CI checks are green now (76 passing, build on latest `main`). Rebased cleanly, and `v1-core-plus-kv-plus-metrics` is green on both NVIDIA and AMD. The earlier failures were main-branch flakes (HF 429 on model download, AMD env) — unrelated to this change.

@ivanium @njhill @zhewenl — the earlier review feedback is all addressed (including moving `resolve_block_hashes` to `kv_cache_utils.py` per @ivanium's suggestion). Could one of you convert your LGTM to a formal approval so this can merge? Happy to address anything else. Thanks!
